File: artifacts/admin-panel/src/adminv2/registry/ribbonAssembly.ts
/**
 * Contextual-tab assembly and the trail that feeds it.
 *
 * The complaint that produced this: "I try to go right back to what I clicked
 * to get where I am and it's a new icon." So the way back is not something each
 * screen draws — it is spliced into every contextual tab at the same position,
 * by this module, and screens cannot opt out.
 */

import { ArrowLeft, Search } from "lucide-react";
import { ACCENT } from "../theme";
import type { ContextualTabSpec, PeekKind, RibbonGroup } from "./types";

/** Where you have been. `openDoc` pushes onto the front. */
export interface TrailEntry {
  /** Peek kind, or "screen" for a plain destination. */
  kind: PeekKind | "screen";
  id: string;
  /** Resolved human name — never a raw id. */
  label: string;
  /** Reopens this entry. */
  open: () => void;
}

/** handoff.md: `state.trail` is capped at 6 and deduped. */
export const TRAIL_MAX = 6;

/**
 * Pushes an entry onto the trail, most-recent-first.
 *
 * Deduped on `kind:id`, so revisiting something moves it to the front rather
 * than growing a run of the same record. Returns a new array; never mutates.
 */
export function pushTrail(trail: readonly TrailEntry[], entry: TrailEntry): TrailEntry[] {
  const key = trailKey(entry);
  const without = trail.filter((t) => trailKey(t) !== key);
  return [entry, ...without].slice(0, TRAIL_MAX);
}

/** `kind:id`, the identity used to dedupe the trail and to exclude the current doc. */
export function trailKey(entry: { kind: PeekKind | "screen"; id: string }): string {
  return `${entry.kind}:${entry.id}`;
}

/**
 * Builds the Back group: the three most recent places that are not where you
 * are now, then "Search everything".
 *
 * `currentKey` is excluded explicitly rather than assumed to be `trail[0]`.
 * That assumption held only while `openDoc` was the sole way focus moved —
 * clicking a doc tab or closing one changes the active doc without touching
 * the trail, and the large button would then offer to navigate to the record
 * already on screen. The design guards it the same way, by filtering the
 * selected doc out before taking the first entry.
 *
 * When there is nothing to go back to, the group still renders with only
 * "Search everything". Hiding it would move every other group one position
 * left, which is exactly the instability this design exists to prevent.
 */
export function backGroupFrom(
  trail: readonly TrailEntry[],
  onSearchEverything: () => void,
  currentKey?: string,
): RibbonGroup {
  const previous = trail.filter((entry) => trailKey(entry) !== currentKey).slice(0, 3);
  const [first, ...rest] = previous;

  return {
    label: "Back",
    large: first
      ? [
          {
            label: first.label,
            icon: ArrowLeft,
            intent: "open",
            onSelect: first.open,
            title: `Back to ${first.label}`,
          },
        ]
      : [],
    small: [
      ...rest.map((entry) => ({
        label: entry.label,
        icon: ArrowLeft as typeof ArrowLeft,
        intent: "open" as const,
        onSelect: entry.open,
        title: `Back to ${entry.label}`,
      })),
      {
        label: "Search everything",
        icon: Search,
        intent: "open" as const,
        onSelect: onSearchEverything,
        color: ACCENT.info,
        title: "Search everything (Ctrl K)",
      },
    ],
  };
}

/** The Back group is spliced in here — second, always. */
export const BACK_GROUP_POSITION = 1;

/**
 * Assembles a contextual tab's final group list.
 *
 * This is the only place a contextual tab's groups are put together. Screens
 * hand over their own groups and never see the Back group; splicing it here is
 * what guarantees the trail sits in the same place on Endpoint Tools, Script
 * Tools, Pipeline Tools, Observe Tools and everything built later.
 *
 * If a screen supplies no groups at all, Back still lands at index 0 rather
 * than being dropped — the group must exist for the tab to be usable.
 */
export function assembleContextualGroups(
  spec: ContextualTabSpec,
  trail: readonly TrailEntry[],
  onSearchEverything: () => void,
  currentKey?: string,
): RibbonGroup[] {
  const groups = [...spec.groups];
  const back = backGroupFrom(trail, onSearchEverything, currentKey);
  const at = Math.min(BACK_GROUP_POSITION, groups.length);
  groups.splice(at, 0, back);
  return groups;
}

File: desktop/MyArchitect/Shell/RibbonAssembly.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace MyArchitect.Shell;

/// <summary>
/// Trail + Back-group injection + fixed-tab group merging — ported from AdminV2's
/// <c>registry/ribbonAssembly.ts</c> (see <c>artifacts/admin-panel/src/adminv2/SHELL.md</c> §2).
/// A screen/Feature never hand-authors a Back group; the shell always splices one in.
/// </summary>
public static class RibbonAssembly
{
    private const int TrailCap = 6;

    /// <summary>Most-recent-first, deduped on Kind:Id, capped at <see cref="TrailCap"/>.</summary>
    public static List<TrailEntry> PushTrail(List<TrailEntry> trail, TrailEntry entry)
    {
        var next = new List<TrailEntry> { entry };
        next.AddRange(trail.Where(t => !(t.Kind == entry.Kind && t.Id == entry.Id)));
        if (next.Count > TrailCap)
        {
            next.RemoveRange(TrailCap, next.Count - TrailCap);
        }
        return next;
    }

    /// <summary>
    /// Builds the Back group: <c>trail[0]</c> is where you *are*, so the large button is
    /// <c>trail[1]</c> (wherever you just came from); below it the two before that; then
    /// "Search everything". Never hidden — when there's nowhere to go back to, the group still
    /// renders with only Search, because hiding it would shift every other group one position
    /// left (the exact instability this exists to prevent).
    /// </summary>
    public static RibbonGroupSpec BackGroupFrom(List<TrailEntry> trail, Action onSearchEverything)
    {
        var group = new RibbonGroupSpec { Label = "Back", Order = -1 };

        if (trail.Count > 1)
        {
            var backTo = trail[1];
            group.Large.Add(new RibbonCommandSpec
            {
                Label = backTo.Label,
                Intent = RibbonIntent.Open,
                OnSelect = backTo.Open,
            });

            foreach (var t in trail.Skip(2).Take(2))
            {
                group.Small.Add(new RibbonCommandSpec
                {
                    Label = t.Label,
                    Intent = RibbonIntent.Open,
                    OnSelect = t.Open,
                });
            }
        }

        group.Small.Add(new RibbonCommandSpec
        {
            Label = "Search everything",
            Intent = RibbonIntent.Open,
            OnSelect = onSearchEverything,
        });

        return group;
    }

    /// <summary>
    /// Splices the Back group at position 2 (index 1) of a contextual tab's own groups — never
    /// position 0, so a contextual tab's own most-important group can still lead.
    /// </summary>
    public static List<RibbonGroupSpec> AssembleContextualGroups(
        ContextualTabSpec spec, List<TrailEntry> trail, Action onSearchEverything)
    {
        var back = BackGroupFrom(trail, onSearchEverything);
        var result = new List<RibbonGroupSpec>();
        var own = spec.Groups.OrderBy(g => g.Order).ToList();

        if (own.Count > 0) result.Add(own[0]);
        result.Add(back);
        result.AddRange(own.Skip(1));
        return result;
    }

    /// <summary>
    /// Merges groups that share a <see cref="RibbonGroupSpec.Label"/> on the same fixed tab into
    /// one physical box (UI_RULES.md §2 — Watch is fed by Alerts #3483, SLA breaches #3487 and
    /// the Task Queue #3490; each gets its own alert but they render as one themed box, not three
    /// tabs). The merged box sorts at whichever contributor asked for the lowest Order.
    /// </summary>
    public static List<RibbonGroupSpec> MergeGroupsByLabel(IEnumerable<RibbonGroupSpec> groups)
    {
        var merged = new Dictionary<string, RibbonGroupSpec>(StringComparer.Ordinal);
        var order = new List<string>();

        foreach (var g in groups)
        {
            if (!merged.TryGetValue(g.Label, out var existing))
            {
                merged[g.Label] = new RibbonGroupSpec { Label = g.Label, Order = g.Order };
                order.Add(g.Label);
                existing = merged[g.Label];
            }
            else if (g.Order < existing.Order)
            {
                // Re-key with the lower order while keeping the same accumulated command lists.
                merged[g.Label] = new RibbonGroupSpec { Label = g.Label, Order = g.Order, Large = existing.Large, Small = existing.Small };
                existing = merged[g.Label];
            }

            existing.Large.AddRange(g.Large);
            existing.Small.AddRange(g.Small);
        }

        return order.Select(l => merged[l]).OrderBy(g => g.Order).ToList();
    }
}
